ami/trainers/i_jepa_latent_visualization_trainer.py

```diff
@@ -0,0 +1,217 @@
+
+    @override
+    def is_trainable(self) -> bool:
+        self.image_data_user.update()
+        return len(self.image_data_user.buffer) >= self.minimum_dataset_size and self._is_new_data_available()
+
+    def _is_new_data_available(self) -> bool:
+        return (
+            self.image_data_user.buffer.count_data_added_since(self.dataset_previous_get_time)
+            >= self.minimum_new_data_count
+        )
+
+    def get_dataset(self) -> Dataset[Tensor]:
+        dataset = self.image_data_user.get_dataset()
+        self.dataset_previous_get_time = time.time()
+        return dataset
+
+    @torch.no_grad()
+    def log_visualization(self, dataloader: DataLoader[Tensor]) -> None:
+        """logs grid of input and reconstructed images for visualization.
+
+        Args:
+            dataloader (DataLoader[Tensor]): DataLoader containing the image batches.
+        """
+        image_batches = []
+        reconstruction_image_batches = []
+        num_remaining = self.num_visualize_images
+
+        batch: tuple[Tensor]
+        for batch in dataloader:
+            (image_batch,) = batch
+            image_batch = image_batch[:num_remaining]
+            image_batches.append(image_batch)
+            image_batch = image_batch.to(self.device)
+
+            latents = self.encoder(image_batch)
+            reconstruction: Tensor = self.decoder(latents)
+
+            reconstruction_image_batches.append(reconstruction.cpu())
+
+            num_remaining -= reconstruction.size(0)
+            if num_remaining <= 0:
+                break
+
+        grid_input_image = torchvision.utils.make_grid(torch.cat(image_batches), self.visualize_grid_row)
+        grid_reconstruction_image = torchvision.utils.make_grid(
+            torch.cat(reconstruction_image_batches), self.visualize_grid_row
+        )
+
+        self.logger.tensorboard.add_image(self.log_prefix + "metrics/input", grid_input_image, self.logger.global_step)
+        self.logger.tensorboard.add_image(
+            self.log_prefix + "metrics/reconstruction", grid_reconstruction_image, self.logger.global_step
+        )
+
+    @override
+    def train(self) -> None:
+        # move to device
+        self.encoder.to(self.device)
+        self.decoder.to(self.device)
+
+        optimizer = self.partial_optimizer(self.decoder.parameters())
+        optimizer.load_state_dict(self.optimizer_state)
+
+        # prepare about dataset
+        dataloader = self.partial_dataloader(dataset=self.get_dataset())
+
+        for _ in range(self.max_epochs):
+            batch: tuple[
+                Tensor,
+            ]
+            for batch in dataloader:
+                (image_batch,) = batch
+                image_batch = image_batch.to(self.device)
+
+                with torch.no_grad():
+                    latents = self.encoder(image_batch)
+                    # latents: [batch_size, n_patches_height * n_patches_width, latents_dim]
```

Comment:
Target EncoderのLatentを可視化したい時は一度`torch.nn.functional.layer_norm`での正規化を挟みたい　学習中の条件と可視化中の条件が変わるのよくない気がする
```suggestion
                with torch.no_grad():
                    latents = self.encoder(image_batch)
                    # latents: [batch_size, n_patches_height * n_patches_width, latents_dim]
                    if target_encoder:
                         latents = torch.nn.functional.layer_norm(latents, (latents.size(-1),))  # normalize over feature-dim
```

Reply:
Target EncoderはForward Dynamicsへ出力される際は Layer Normされないんだ。だからこれでいい

Reply:
ん、でもどっちでもいいといえばいいか。

Reply:
それForward Dynamicsに渡す時もLayer Norm必要なやつなのでは

Reply:
正規化って結構影響でそうだけどなあ　例えばBatchNormのあるなしでGANとかの学習結果って全然変わるの有名だし

Reply:
一応 https://github.com/MLShukai/ami-vconf24/pull/100/commits/6fd50d3f274ab8368bad96f1df61050d5fcfc28a でTarget Encoder指定時にのみ正規化する処理を追加しておいた
もし不要だと思うなら`git revert`なり何なりでこのコミット無かったことにしといて

Reply:
@zassou65535 その通り過ぎたので `i_jepa_encoder_infer`を修正した。これでforward dynamicsでも layer normが適用される。https://github.com/MLShukai/ami-vconf24/pull/100/commits/66cab38a96bed1847db2d5e173cf966c3ee2c845

Reply:
Context Encoderでも必要かもしれない。まだ完全にContextとTarget, どちらをForward Dynamicsへ用いるかは定まっていないからね。